File: Packages/Global/Apache/RestServer/RestServer.php
<?php

// 13.04.2022


namespace Apache;

require_once __dir__ . '/../Json/Json.php';


class RestServer {
    static public $errorTrace = false;
    static public $gzip_level = 0;
    static public $request_method = '';
    static public $timeLimit = 60;


    public $_data = [];
    public $_method = '';
    public $_method_args = [];
    public $_timeStamp = 0;


    public function _init() {}

    public function _request__parse() {
        if (static::$request_method && $_SERVER['REQUEST_METHOD'] != static::$request_method) {
            throw new \Exception('Request_method');
        }

        if ($_SERVER['REQUEST_METHOD'] == 'POST') {
            $request_body = file_get_contents('php://input');
            $request_data = Json::parse($request_body);

            $this->_data = $request_data['data'] ?? [];
            $this->_method = $request_data['method'] ?? '';
            $this->_method_args = $request_data['method_args'] ?? [];
        }
        else {
            $this->_method = $_GET['method'] ?? '';
            $this->_method_args = $_GET['method_args'] ?? [];
        }

        if (!$this->_method || str_starts_with($this->_method, '_')) {
            throw new \Exception('Method');
        }
    }

    public function _run() {
        $result = null;

        try {
            $this->_request__parse();
            $this->_init();

            $result = $this->{$this->_method}(...$this->_method_args);
            $result = ['result' => $result];
        }
        catch (\Error $error) {
            $result = ['error' => $error->getMessage()];

            if (static::$errorTrace) {
                $result['errorTrace'] = $error->getTrace();
            }
        }
        catch (\Exception $exception) {
            $result = ['exception' => $exception->getMessage()];

            if (static::$errorTrace) {
                $result['errorTrace'] = $exception->getTrace();
            }
        }

        $result = Json::stringify($result);

        if (static::$gzip_level) {
            $result = gzEncode($result, static::$gzip_level);
            header('Content-Encoding: gzip');
        }

        echo $result;
    }

    public function _timeLimit__check() {
        return microTime(true) - $this->_timeStamp <= static::$timeLimit;
    }


    public function __construct() {
        set_time_limit(0);

        $this->_timeStamp = microTime(true);
        $this->_run();
    }
}

File: Packages/Global/Python/RestServer/RestServer.py
# 15.08.2024

import cgi
import gzip
import time
import traceback
import sys

from HttpServer.HttpServer import HttpServer
from Json.Json import Json


class RestServer(HttpServer):
    error_trace = False
    gzip_level = 0
    request_method_server = ''
    time_limit = 60


    def _init(self): pass

    def _request__parse(self):
        if self.request_method_server and self.request_method != self.request_method_server:
            raise Exception('Request_method')

        request_data = self._request_data__get()
        self._data = request_data
        # self._data = request_data.get('data', [])
        self._method = request_data.get('method', '')
        self._method_args = request_data.get('method_args', [])
        self._method_args = Json.parse(self._method_args) if isinstance(self._method_args, str) else self._method_args



        if not self._method or self._method.startswith('_'):
            raise Exception('Method')

    def _run(self):
        result = None
        self._time_stamp = time.time()

        try:
            self._request__parse()
            self._init()

            method = getattr(self, self._method, None)

            if callable(method):
                result = method(*self._method_args)
            else:
                raise Exception('Method')

            result = {'result': result}
        except Exception as exception:
            result = {'exception': str(exception)}

            if self.error_trace:
                result['errorTrace'] = traceback.format_exc()

        if self.gzip_level:
            result = gzip.compress(Json.stringify(result).encode(), self.gzip_level)
            self.send_header('Content-Encoding', 'gzip')

        self.echo(result)


    def do_GET(self):
        self._run()

    def do_POST(self):
        super().do_POST()
        self._run()


    def __init__(self, *args):
        super().__init__(*args)

        self._data = []
        self._method = ''
        self._method_args = []
        self._time_stamp = 0

